README.md
# ForeComp

This package generates a size-power tradeoff plot for a given Diebold-Mariano test.

## Install latest version from GitHub
```R
install.packages("devtools")
devtools::install_github("mcmcs/Forecomp", ref = "19-submit-to-cran")
```




NEWS.md
# ForeComp

# ForeComp 0.9.0 (1 August 2023)
  * First public release.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/testing-design.html#sec-tests-files-overview
# * https://testthat.r-lib.org/articles/special-files.html

library(testthat)
library(ForeComp)

test_check("ForeComp")


tests/testthat/test-Plot_Tradeoff.R


# =====================================================
# Test 1

test_that("Size distortion cannot be smaller than -5%, which is the pre-specified confidence level for this package", {


  expect_true(min(Plot_Tradeoff(data = TBILL,
                                f1   = "SPFfor_Step1",
                                f2   = "NCfor_Step1",
                                y    = "Realiz1")[[2]]$b_size_distortion) >= -0.05)

})

# =====================================================
# Test 2
test_that("Check the dimension of the output.", {


  m_set = c(1,2,3,4,5);

  output = Plot_Tradeoff(data = TBILL,
                f1   = "SPFfor_Step1",
                f2   = "NCfor_Step1",
                y    = "Realiz1",
                m_set = m_set);

  expect_true(nrow(output[[2]]) == length(m_set));

})

# =====================================================
# Test 3
test_that("Check that the M labels are plotted by default.", {

  output = Plot_Tradeoff(data = TBILL,
                f1   = "SPFfor_Step1",
                f2   = "NCfor_Step1",
                y    = "Realiz1"
                );

  expect_false(is.null(output[[1]]$layers[[3]][["constructor"]][[2]][["label"]]))
})

# =====================================================
# Test 4
test_that("Check points have no labels when no_m_label = TRUE.", {

  output = Plot_Tradeoff(data = TBILL,
                f1   = "SPFfor_Step1",
                f2   = "NCfor_Step1",
                y    = "Realiz1",
                no_m_label = TRUE
                );

  expect_error(is.null(output[[1]]$layers[[3]]$computed_mapping$label))

})


R/data.R

#' Civilian Unemployment Rate (UNEMP)
#'
#' Error Statistics for the Survey of Professional Forecasters for Unemployment Rate. All NAs are replaced with 0 as in Coroneo and Iacone (2020).
#'
#' @format `UNEMP`
#' A data frame with 219 rows and 16 columns
#' @references Coroneo, L., Iacone, F. (2020), Comparing Predictive Accuracy in Small Samples Using Fixed-Smoothing Asymptotics, \emph{Journal of Applied Econometrics}, \bold{35}(4), 391-409.
#' @source <https://www.philadelphiafed.org/-/media/frbp/assets/surveys-and-data/survey-of-professional-forecasters/data-files/unemp/data_spf_error_statistics_unemp_1_aic.xls?la=en&hash=4CAD0B11FEAB6C4D0F30C38965FE3354>
"UNEMP"


#' 3-Month Treasury Bill Rate (TBILL)
#'
#' Error Statistics for the Survey of Professional Forecasters for Treasury Bill Rate (Three Month). All NAs are replaced with 0 as in Coroneo and Iacone (2020).
#'
#' @format `TBILL`
#' A data frame with 219 rows and 16 columns
#' @references Coroneo, L., Iacone, F. (2020), Comparing Predictive Accuracy in Small Samples Using Fixed-Smoothing Asymptotics, \emph{Journal of Applied Econometrics}, \bold{35}(4), 391-409.
#' @source <https://www.philadelphiafed.org/-/media/frbp/assets/surveys-and-data/survey-of-professional-forecasters/data-files/tbill/data_spf_error_statistics_tbill_1_aic.xls?la=en&hash=F432350F84B0E4CCE9A1E9D399447CA9>
"TBILL"


#' Real Gross National Product/Gross Domestic Product (RGDP)
#'
#' Error Statistics for the Survey of Professional Forecasters for Real GNP/GDP. All NAs are replaced with 0 as in Coroneo and Iacone (2020).
#'
#' @format `RDGP`
#' A data frame with 219 rows and 16 columns
#' @references Coroneo, L., Iacone, F. (2020), Comparing Predictive Accuracy in Small Samples Using Fixed-Smoothing Asymptotics, \emph{Journal of Applied Econometrics}, \bold{35}(4), 391-409.
#' @source <https://www.philadelphiafed.org/-/media/frbp/assets/surveys-and-data/survey-of-professional-forecasters/data-files/rgdp/data_spf_error_statistics_rgdp_3_aic.xls?la=en&hash=3AC9E2D8A5299F93CA7E16CFAA974C22>
"RGDP"


#' Price Index for Gross National Product/Gross Domestic Product (PGDP)
#'
#' Error Statistics for the Survey of Professional Forecasters for GNP/GDP Price Index. All NAs are replaced with 0 as in Coroneo and Iacone (2020).
#'
#' @format `PGDP`
#' A data frame with 219 rows and 16 columns
#' @references Coroneo, L., Iacone, F. (2020), Comparing Predictive Accuracy in Small Samples Using Fixed-Smoothing Asymptotics, \emph{Journal of Applied Econometrics}, \bold{35}(4), 391-409.
#' @source <https://www.philadelphiafed.org/-/media/frbp/assets/surveys-and-data/survey-of-professional-forecasters/data-files/pgdp/data_spf_error_statistics_pgdp_3_aic.xls?la=en&hash=148987D03D54DA5391A44F28CBC12729>
"PGDP"



R/forecast_eval_functions.R
# Functions to perform tests
# 9/15/2019

## To build a documentation (in help window)
## devtools::document()

## To build a manual (in pdf)
## devtools::build_manual(path=getwd())

# ----------------------------------------------------------------------
# Two-sided tests
# ----------------------------------------------------------------------


#' Diebold-Mariano Test (with an original recommendation)
#'
#' This function performs the Diebold-Mariano test with their original recommendation.
#' Let \eqn{d_{t}}{d(t)} be a sequence of loss differential, \eqn{t=1,2,...,T}.
#' Then, the function performs a statistical test for the following null hypothesis:
#' \deqn{E[d_{t}] = 0}{E[d(t)] = 0}
#' This function follows the original recommendation by Diebold and Mariano (1995),
#' where the long-run variance is estimated using the rectangular kernel truncated at
#' \eqn{(h-1)}. This function performs a two-sided test. Normal approximation is used to calculate critical values.
#' @param d loss differential
#' @param h h-step-ahead forecast (default = 1, i.e., comparing one-step-ahead forecasts)
#' @param cl confidence level (default = 0.05, i.e., 5\%)
#' @return This function returns a class with the following elements
#' \itemize{
#' \item \code{rej} is a T/F value. TRUE (reject), FALSE (accept)
#' \item \code{stat} is a test statistic
#' \item \code{pval} is an associated p-value
#' }
#' @author Minchul Shin
#' @keywords internal
#' @noRd


dm.test.r = function(d, h=1, cl = 0.05){

  n = length(d);
  d.cov = stats::acf(d, lag.max = (h-1), type="covariance", plot=FALSE)$acf[,,1];
  d.var = sum(c(d.cov[1], 2*d.cov[-1]))/n;


  dmstat = mean(d)/sqrt(d.var);

  pval = 2 * stats::pnorm(-abs(dmstat), mean=0, sd=1); #p-val

  rej = pval < cl; #(cl)-level testing, rej-decision

  if (d.var<0){
    warning("dm.eq: negative variance ... ");
    rej = NA; #negative variance
  }

  #output
  outls = list();
  outls$rej = rej;
  outls$stat = dmstat;
  outls$pval = pval;

  return(outls);
}


#' Diebold-Mariano Test (Modified-DM)
#'
#' Diebold-Mariano Test (Modified-DM). Finite-sample modification to the original DM's test.
#' This is a two-sided test.
#'
#' This function is based on \link[forecast]{dm.test} in the "forecast" package on CRAN.
#'
#' @param d loss differential
#' @param h h-step-ahead forecast (default = 1, i.e., one-step-ahead forecasts)
#' @param cl confidence level (default = 0.05, i.e., 5\%)
#' @return This function returns a class with the following elements
#' \itemize{
#' \item \code{rej} is a T/F value. TRUE (reject), FALSE (accept)
#' \item \code{stat} is a test statistic
#' \item \code{pval} is an associated p-value
#' }
#' @author Minchul Shin
#' @keywords internal
#' @noRd
dm.test.r.m = function(d, h=1, cl = 0.05){

  n = length(d);
  d.cov = stats::acf(d, lag.max = (h-1), type="covariance", plot=FALSE)$acf[,,1];
  d.var = sum(c(d.cov[1], 2*d.cov[-1]))/n;


  dmstat = mean(d)/sqrt(d.var);

  # Finite-sample modification
  k = ((n + 1 - 2*h + (h/n) * (h - 1))/n)^(1/2);
  dmstat = dmstat * k; #finite sample correction

  # Another finite-sample modification (to use t-dist)
  pval = 2 * stats::pt(-abs(dmstat), df = (n-1) ); #p-val

  rej = pval < cl; #(cl)-level testing, rej-decision

  if (d.var<0){
    warning("dm.eq: negative variance ... ");
    rej = NA; #negative variance
  }

  #output
  outls = list();
  outls$rej = rej;
  outls$stat = dmstat;
  outls$pval = pval;

  return(outls);
}


# ---------------------------------------------------------------
#' Diebold-Mariano Test (Bartlett kernel, normal approximation)
#'
#' Diebold-Mariano Test (Bartlett kernel, normal approximation). This is a two-sided test.
#'
#' @param d loss differential
#' @param M truncation parameter for the Bartlett kernel (if \code{M = NA}, then \code{Mopt = 2} by default)
#' @param Mopt option for optimal bandwidth, 1 if Lazarus et al. (2018), 2 if Newey and West (1994)
#' @param cl confidence level (default = 0.05, i.e., 5\%)
#' @return This function returns a class with the following elements
#' \itemize{
#' \item \code{rej} is a T/F value. TRUE (reject), FALSE (accept)
#' \item \code{stat} is a test statistic
#' \item \code{pval} is an associated p-value
#' }
#' @author Minchul Shin
#' @keywords internal
#' @noRd
dm.test.bt = function(d, M = NA, Mopt = NA, cl = 0.05){

  n = length(d);

  # Bandwidth selection (if M is not specified)
  # For Bartlett Kernel with normal approximation, NW (1994) is a default
  if (is.na(M)){
    if (is.na(Mopt)){
      Mopt = 2; #Newey and West (1994)
    }

    if (Mopt == 1){
      M = floor(1.3*sqrt(n)) + 1; #Lazarus et al. (2018)
    } else if (Mopt == 2) {
      M = floor(4*(n/100)^(2/9)) + 1; #Newey and West (1994);
    }
  }

  # Long-run variance estimation
  d.cov = stats::acf(d, lag.max = M, type="covariance",plot=FALSE)$acf[,,1];
  d.var = ( d.cov[1] + 2* sum( ( 1 - ( (1:M)/M ) ) * d.cov[-1] ) ) / n;

  dmstat = mean(d) / sqrt(d.var);

  pval = 2 * stats::pnorm(-abs(dmstat), mean=0, sd=1); #p-val based on normal approximation

  rej = pval < cl; #reject decision

  if (d.var<0){
    # Note: in principle, it cannot be negative with prob. one.
    warning("dm.eq.bt: negative variance ... ");
    rej = NA; #negative variance
  }

  #output
  outls = list();
  outls$rej = rej;
  outls$stat = dmstat;
  outls$pval = pval;

  return(outls);
}

# ---------------------------------------------------------------
#' Diebold-Mariano Test (Bartlett kernel, fixed-b approximation)
#'
#' Diebold-Mariano Test (Bartlett kernel, fixed-b approximation). This is a two-sided test.
#'
#' @param d loss differential
#' @param M truncation parameter for the Bartlett kernel (if \code{M = NA}, then \code{Mopt = 1} by default)
#' @param Mopt option for optimal bandwidth, 1 if Lazarus et al. (2018), 2 if Newey and West (1994)
#' @param cl confidence level (default = 0.05, i.e., 5\%), Only 0.05 (5\%) or 0.10 (10\%) are allowed.
#' @return This function returns a class with the following elements
#' \itemize{
#' \item \code{rej} is a T/F value. TRUE (reject), FALSE (accept)
#' \item \code{stat} is a test statistic
#' \item \code{pval} is an associated p-value
#' }
#' @author Minchul Shin
#' @keywords internal
#' @noRd
dm.test.bt.fb = function(d, M = NA, Mopt = NA, cl = 0.05){

  n = length(d);

  # Bandwidth selection (if M is not specified)
  # For Bartlett Kernel with fixed-b approximation, Lazarus et al. is a default
  if (is.na(M)){
    if (is.na(Mopt)){
      Mopt = 1; #Lazarus et al. (2018)
    }

    if (Mopt == 1){
      M = floor(1.3*sqrt(n)) + 1; #Lazarus et al. (2018)
    } else if (Mopt == 2) {
      M = floor(4*(n/100)^(2/9)) + 1; #Newey and West (1994);
    }
  }

  # Long-run variance estimation
  d.cov = stats::acf(d, lag.max = M, type="covariance",plot=FALSE)$acf[,,1];
  d.var = ( d.cov[1] + 2* sum( ( 1 - ( (1:M)/M ) ) * d.cov[-1] ) ) / n;

  dmstat = mean(d) / sqrt(d.var);

  # Kiefer and Vogelsang (2005)
  b = M/n;
  #crit950 = 1.6449 + 2.1859*b + 0.3142*b^2 -0.3427*b^3; #0.950 quantile
  #crit975 = 1.9600 + 2.9694*b + 0.4160*b^2 -0.5324*b^3; #0.975 quantile
  if (cl == 0.05){
    crit = 1.9600 + 2.9694*b + 0.4160*b^2 -0.5324*b^3; #0.975 quantile
  } else if (cl == 0.10){
    crit = 1.6449 + 2.1859*b + 0.3142*b^2 -0.3427*b^3; #0.950 quantile
  } else {
    warning("dm.eq.bt.fb: cl has to be 0.05 or 0.10 ... function exited without testing");
    outls = list();
    return(outls);
  }

  # rejection decision
  rej = abs(dmstat) > crit; #reject decision

  if (d.var<0){
    # Note: in principle, it cannot be negative with prob. one.
    warning("dm.eq.bt.fb: negative variance ... ");
    rej = NA; #negative variance
  }

  #output
  outls = list();
  outls$rej = rej;
  outls$stat = dmstat;

  return(outls);
}


# ---------------------------------------------------------------
#' Diebold-Mariano Test (EWC, fixed-b approximation)
#'
#' Diebold-Mariano Test (EWC, fixed-b approximation). This is a two-sided test.
#'
#' @param d loss differential
#' @param B truncation parameter for the EWC long-run variance estimator (if \code{B = NA}, then \code{Bopt = 1} by default)
#' @param Bopt option for optimal bandwidth, 1 if Lazarus et al. (2018)' recommendation
#' @param cl confidence level (default = 0.05, i.e., 5\%)
#' @return This function returns a class with the following elements
#' \itemize{
#' \item \code{rej} is a T/F value. TRUE (reject), FALSE (accept)
#' \item \code{stat} is a test statistic
#' \item \code{pval} is an associated p-value
#' }
#' @author Minchul Shin
#' @keywords internal
#' @noRd
dm.test.ewc.fb = function(d, B = NA, Bopt = NA, cl = 0.05){

  n = length(d);

  # Bandwidth selection (if M is not specified)
  # For Bartlett Kernel with fixed-b approximation, Lazarus et al. is a default
  if (is.na(B)){
    if (is.na(Bopt)){
      Bopt = 1; #Lazarus et al. (2018)
    }

    if (Bopt == 1){
      B = floor(0.4*n^(2/3)); #Lazarus et al. (2018)
    }
  }

  # EWC estimator
  j_ind = matrix(1:B);

  t_ind = ( t(matrix(1:n)) - 1/2 ) / n;

  Lhat = sqrt(2/n) * ( cos(pi*j_ind %*% t_ind) %*% d ); #eqn (10) of Lazarus et al. (2018)
  Ohat = crossprod(Lhat) / B; #eqn (11)

  d.var = Ohat / n;
  dmstat = mean(d) / sqrt(d.var);

  pval = 2 * stats::pt(-abs(dmstat), df = B); #p-val


  rej = pval < cl; #reject decision

  if (d.var<0){
    # Note: in principle, it cannot be negative with prob. one.
    warning("dm.eq.ewc: negative variance ... ");
    rej = NA; #negative variance
  }

  #output
  outls = list();
  outls$rej = rej;
  outls$stat = dmstat;
  outls$pval = pval;

  return(outls);
}


# ---------------------------------------------------------------
#' Diebold-Mariano Test (WPE, fixed-m approximation)
#'
#' Diebold-Mariano Test (WPE, fixed-m approximation). This is a two-sided test. See Coroneo and Iacone (2020)
#'
#' @param d loss differential
#' @param M truncation parameter for the WPE long-run variance estimator (if \code{M = NA}, then \code{Mopt = 1} by default)
#' @param Mopt option for optimal bandwidth, 1 if Coroneo and Iacone's default value (M = floor(T^(1/3)))
#' @param cl confidence level (default = 0.05, i.e., 5\%)
#' @return This function returns a class with the following elements
#' \itemize{
#' \item \code{rej} is a T/F value. TRUE (reject), FALSE (accept)
#' \item \code{stat} is a test statistic
#' \item \code{pval} is an associated p-value
#' }
#' @author Minchul Shin
#' @keywords internal
#' @noRd
dm.test.wpe.fb = function(d, M = NA, Mopt = NA, cl = 0.05){

  n = length(d);

  # Bandwidth selection (if M is not specified)
  if (is.na(M)){
    if (is.na(Mopt)){
      Mopt = 1; # Coroneo and Iacone's default
    }

    if (Mopt == 1){
      M = floor(n^(1/3)); #Lazarus et al. (2018)
    }
  }

  # WPE as in Coroneo and Iacone (2020)
  lamj = matrix(2*pi/n * (1:M));
  gam2 = 2*pi*mean( abs( 1/(sqrt(2*pi*n)) * ( exp( -1i * ( lamj %*% t( matrix(1:n)) ) ) %*% d ) )^2 ); # Eqn(8)
  dmstat = sqrt(n) * mean(d)/sqrt(gam2) #Eqn(9)
  pval = 2 * stats::pt(-abs(dmstat), df = 2*M); #p-val
  rej = pval < cl; #reject decision

  if (gam2<0){
    # Note: in principle, it cannot be negative with prob. one.
    warning("dm.eq.ewc: negative variance ... ");
    rej = NA; #negative variance
  }

  #output
  outls = list();
  outls$rej = rej;
  outls$stat = dmstat;
  outls$pval = pval;

  return(outls);
}

# ---------------------------------------------------------------
#' Ibragimov and Muller (2010)
#'
#' t-Statistic based HAR-inference by Ibragimov and Muller (2010).
#'
#' @param d loss differential
#' @param q number of blocks
#' @param cl confidence level (default = 0.05, i.e., 5\%)
#' @return This function returns a class with the following elements
#' \itemize{
#' \item \code{rej} is a T/F value. TRUE (reject), FALSE (accept)
#' \item \code{stat} is a test statistic
#' \item \code{pval} is an associated p-value
#' }
#' @author Minchul Shin
#' @keywords internal
#' @noRd
dm.test.im = function(d, q = 2, cl = 0.05){

  n = length(d);

  # b = n/q; #block size (q should be multiple of n, it does not have to be, but for an ease of computation ... )

  # poorman's way of getting out from the case in which n is not a multiple of q
  b = ceiling(n/q);
  D = rep(NA,(b*q));
  D[1:n] = d;

  m = apply(matrix(D, nrow = b, ncol = q), 2, mean, na.rm=T);

  # eqn (17) in CRS
  mbar   = mean(m);
  s2bar  = sum((m-mbar)^2)/(q-1);
  dmstat = mbar / sqrt(s2bar/q);

  # critical value from t-distribution
  pval = 2 * stats::pt(-abs(dmstat), df = (q-1)); #p-val
  rej = pval < cl; #reject decision

  #output
  outls = list();
  outls$rej  = rej;
  outls$stat = dmstat;
  outls$pval = pval;

  return(outls);
}



R/Plot_Tradeoff.R

#' Visualizes the size distortion maximum power loss tradeoff from the Diebold-Mariano test for equal predictive accuracy
#'
#' @description `Plot_Tradeoff` creates a plot to show sensitivity of statistical significance to the choice of bandwidth and how size distortion and maximum power loss vary.
#' It is designed for the Diebold-Mariano test for equal predictive accuracy (Diebold and Mariano, 2002).
#' For a size-power tradeoff plot, see Lazarus, Lewis, Stock, and Watson (2018) and Lazarus, Lewis, and Stock (2021).
#'
#' @param data A data frame.
#' @param f1 Column containing forecaster 1's predictions. Should be a string.
#' @param f2 Column containing forecaster 2's predictions. Should be a string.
#' @param y Column containing the realized value for the outcome variable. Should be a string.
#' @param loss_function The transformation applied to the forecast error. Defaults to squared error loss. The user supplied function should take two inputs and a scalar output, loss = loss_function(f, y). For example, quadratic loss function would be defined as \code{loss_function=function(f,y){(f-y)^2}}.
#' @param n_sim The number of simulations used to generate the ARIMA model. Defaults to 1,000.
#' @param m_set The truncation parameter. Defaults to c(1:10, seq( 11, floor(nrow(data)/2), 10)). For a standard long-run variance calculation (for example, using Bartlett kernel), it controls the number of terms used in estimating the autocovariance matrix. It should be a vector of integers with the values of M you would like to plot.
#' @param verbose TRUE to print out the progress to the console. Defaults to TRUE.
#' @param no_m_label TRUE to plot without m labels. Defaults to FALSE.
#' @returns A list of length 2. The first element is a ggplot2 object of the size-power tradeoff. The second element is the underlying data used to construct the plot in element 1.
#' @author Nathan Schor and Minchul Shin
#' @importFrom forecast auto.arima
#' @importFrom stats acf
#' @importFrom stats pnorm
#' @importFrom stats approx
#' @importFrom stats arima.sim
#' @importFrom stats quantile
#' @importFrom stats rnorm
#' @importFrom astsa arma.spec
#' @importFrom rlang .data
#' @import ggplot2
#' @export
#' @references Diebold, F. X. & Mariano, R. S. (2002), Comparing Predictive Accuracy, \emph{Journal of Business & Economic Statistics}, \bold{20}(1), 134-144.
#' @references Lazarus, E., Lewis, D. J., Stock, J. H. & Watson, M. W. (2018), HAR Inference: Recommendations for Practice, \emph{Journal of Business & Economic Statistics}, \bold{36}(4), 541-559.
#' @references Lazarus, E., Lewis, D. J. & Stock, J. H. (2021), The Size-Power Tradeoff in HAR Inference, \emph{Econometrica}, \bold{89}(5), 2497-2516.
#' @examples
#'
#' # A typical example
#' set.seed(1234)
#' output = Plot_Tradeoff(
#'   data = TBILL,
#'   f1   = "SPFfor_Step1",
#'   f2   = "NCfor_Step1",
#'   y    = "Realiz1",
#'   m_set = seq(from = 1, to = 70, by = 10)
#' )
#' output[[1]] # The first element is a ggplot2 object of the size-power tradeoff.
#' output[[2]] # The second element is the underlying data used to construct the plot in element 1.
#'
#' # An example with a user supplied loss function
#' # To use the mean absolute error as a loss function rather than a quadratic loss function
#' set.seed(1234)
#' output = Plot_Tradeoff(
#'   data = TBILL,
#'   f1   = "SPFfor_Step1",
#'   f2   = "NCfor_Step1",
#'   y    = "Realiz1",
#'   loss_function = function(f,y){ abs(f-y) },
#'   m_set = seq(from = 1, to = 50, by = 10)
#' )
#'
#' # An example without (f1, f2, y). The function will take the first three columns and use them
#' set.seed(1234)
#' tmpdata = TBILL[, c("SPFfor_Step1", "NCfor_Step1", "Realiz1")] # data with [f1, f2, y]
#' Plot_Tradeoff(
#'   data = tmpdata,
#'   m_set = seq(from = 1, to = 50, by = 10)
#' )
#'



Plot_Tradeoff <- function(data,
                          f1 = NULL,
                          f2 = NULL,
                          y  = NULL,
                          loss_function = NULL,
                          n_sim = 1000,
                          m_set = NULL,
                          verbose = TRUE,
                          no_m_label = FALSE) {

  # Handling options

  # Data
  if (is.null(f1)& is.null(f2) & is.null(y)){
    f1 = data[[1]];
    f2 = data[[2]];
    y  = data[[3]];
  } else if (!is.null(f1) & !is.null(f2) & !is.null(y)) {
    f1 <- data[[f1]]
    f2 <- data[[f2]]
    y <- data[[y]]
  } else {
    stop("f1, f2, y have to be supplied altogether.")
  }

  # If the user does not supply a loss_function, we use a quadratic loss
  if (is.null(loss_function)) {
    loss_function = function(f, y){ return( (f-y)^2 ); };
  }

  # calculating loss
  loss1 = loss_function(f1, y);
  loss2 = loss_function(f2, y);
  d     = loss1 - loss2;

  # mset - default value
  if (is.null(m_set)){
    if ( floor(nrow(data)/2) > 10){
      m_set = c(1:10, seq( 11, floor(nrow(data)/2), 10));
    } else {
      m_set = seq(1, floor(nrow(data)/2), 1);
    }
  }

  if (!is.logical(no_m_label)) {stop("Argument 'no_m_label' should be either TRUE or FALSE.")}
  if (!is.numeric(n_sim) | n_sim %% 1 != 0 | n_sim <= 0) {stop("Argument 'n_sim' should be a natural number.")}


  # other info
  conf_level <- 0.05
  series_length <- nrow(data)
  m_set_length <- length(m_set)

  # matrix to store
  mat_size_distortion_dm <- matrix(NA, m_set_length, 1)
  mat_power_loss_dm <- matrix(NA, m_set_length, 1)

  mat_size_distortion_b <- matrix(NA, m_set_length, 1)
  mat_power_loss_b <- matrix(NA, m_set_length, 1)

  mat_size_distortion_d <- matrix(NA, m_set_length, 1)
  mat_power_loss_d <- matrix(NA, m_set_length, 1)

  mut <- mean(d); # demean
  dt_tilde <- d - mut


  # --- Estimate ARIMA and extract information
  a = forecast::auto.arima(y=dt_tilde, max.p = 12, max.q=12, stationary=T, ic="aic", seasonal=F, allowmean=F);
  i_ar = grep("ar", names(a$coef));
  i_ma = grep("ma", names(a$coef));
  m_sim = list("ar"=a$coef[i_ar], "ma"=a$coef[i_ma]);

  if (length(m_sim$ar)==0){m_sim$ar=0.0}
  if (length(m_sim$ma)==0){m_sim$ma=0.0}

  # --- Generate data for size and power computation (this loop alos pre-computes long-run variance)
  Mmax = min(c(series_length-1,200)); # maximum possible M considered in this experimen
  mat_dt   = matrix(NA, series_length, n_sim); # matrix that stores data (for size calculation)
  mat_dtm  = matrix(NA, 1, n_sim);    # matrix that stores mean of data
  mat_acf  = matrix(NA, (Mmax+1), n_sim); # matrix that stores acf
  for (irep in 1:n_sim){

    # data
    dt_sim = arima.sim(m_sim, n=series_length, innov = rnorm(series_length, 0, sqrt(a$sigma2)));
    mat_dt[,irep] = dt_sim;
    mat_dtm[,irep] = mean(dt_sim);
    # autocovariance matrix
    d.cov = stats::acf(dt_sim, lag.max = (Mmax), type="covariance", plot=FALSE, demean=TRUE)$acf[,,1];
    mat_acf[,irep] = d.cov;
  }


  # --- Setting for trade-off figure
  del_grid = seq(from=0, to=10, by=0.25);
  ndel = length(del_grid);

  v_M <- vector(mode = "integer", length = m_set_length)
  v_hypothesis_test_b <- vector(mode = "logical", length = m_set_length)
  v_test_statistic_b <- vector(mode = "logical", length = m_set_length)
  v_hypothesis_test_dm <- vector(mode = "logical", length = m_set_length)
  v_test_statistic_dm <- vector(mode = "logical", length = m_set_length)

  # --- Loop over M set
  for (iM in 1:m_set_length){

    Mchoice = m_set[iM]; #our choice of M for this iteration

    testres_b  = dm.test.bt.fb(d, cl = conf_level, M = Mchoice);
    wce_b_rej  = testres_b$rej
    wce_b_stat = testres_b$stat

    testres_dm  = dm.test.bt(d, cl = conf_level, M = Mchoice);
    wce_dm_rej  = testres_dm$rej  # CHANGE
    wce_dm_stat = testres_dm$stat # CHANGE

    v_M[iM] <- Mchoice
    v_hypothesis_test_b[iM] <- wce_b_rej
    v_test_statistic_b[iM] <- wce_b_stat
    v_hypothesis_test_dm[iM] <- wce_dm_rej
    v_test_statistic_dm[iM] <- wce_dm_stat

    # --- Oracle ---

    # --- Power for oracle

    # long-run variance
    ss = astsa::arma.spec(ar = m_sim$ar, ma = m_sim$ma, var.noise = a$sigma2, n.freq = 100);
    Om = ss$spec[1]; #this is 2*pi*f(0), spectrum at zero rather than a spectral density at zero

    # --- Standard test


    # --- Size computation for DM-WCE-dm, DM-WCE-b, DM-WPE-d
    M = Mchoice;
    mat_stat =    matrix(NA, n_sim, 1);
    mat_d.var   = matrix(NA, n_sim, 1);
    mat_rej_dm  = matrix(NA, n_sim, 1);
    mat_rej_b   = matrix(NA, n_sim, 1);
    for (irep in 1:n_sim){
      d.cov  = mat_acf[1:(M+1),irep];
      d.var  = ( d.cov[1] + 2*sum( (1 - ((1:M)/M) ) * d.cov[-1] ) ) / series_length;

      # Traditional NW, DM (DM-WCE-dm)
      dmstat = mean(mat_dtm[,irep]) / sqrt(d.var);
      pval   = 2 * stats::pnorm(-abs(dmstat), mean=0, sd=1); #p-val based on normal approximation
      rej_dm    = pval < conf_level; #reject decision

      # Fixed-b NW (DM-WCE-b)
      b = M / series_length;
      if (conf_level == 0.05){
        crit = 1.9600 + 2.9694*b + 0.4160*b^2 -0.5324*b^3; #0.975 quantile
      } else if (conf_level == 0.10){
        crit = 1.6449 + 2.1859*b + 0.3142*b^2 -0.3427*b^3; #0.950 quantile
      }
      # rejection decision
      rej_b = abs(dmstat) > crit; #reject decision


      # ***Minchul DM-WPE-d
      # do be done


      # store results
      mat_stat[irep,] = dmstat;
      mat_d.var[irep,] = d.var;
      mat_rej_dm[irep, ] = rej_dm;
      mat_rej_b[irep, ] = rej_b;
    }
    size_distortion_dm = mean(mat_rej_dm) - conf_level;
    size_distortion_b = mean(mat_rej_b) - conf_level;


    # --- Power of a standard test (DM-NW)
    # Note that WCE-DM and WCE-B have the same power property when size-corrected
    # --- size-corrected crit val
    c05_star = quantile(abs(mat_stat), (1-conf_level));

    # --- size-corrected power
    mat_stat = matrix(NA, n_sim, ndel);
    mat_rej  = matrix(NA, n_sim, ndel);
    mat_rej2 = matrix(NA, n_sim, ndel);
    for (irep in 1:n_sim){

      m_i = mat_dtm[,irep];
      s_i = sqrt(mat_d.var[irep,]);

      for (idel in 1:ndel){

        # DM test with NW
        dmstat = (m_i + (1/sqrt(series_length)) *sqrt(Om)*del_grid[idel]) / s_i; #statistic
        pval = 2 * stats::pnorm(-abs(dmstat), mean=0, sd=1); #p-val based on normal approximation
        rej = pval < conf_level; #reject decision

        mat_rej[irep, idel] = rej; #for raw power
        mat_stat[irep, idel] = dmstat;
        mat_rej2[irep, idel] = abs(dmstat) > c05_star; # for size-corrected power
      }
    }


    #======================================

    # --- another Oracle ---
    grid = seq(from=0, to=5, by=0.05);
    samp = del_grid;
    pow_gau = stats::pnorm(-1.96+grid) + stats::pnorm(-1.96-grid);

    # --- Maximum power loss
    powinterp = stats::approx(samp, apply(mat_rej2, 2, mean), grid)$y; #interpolated power
    max_power_loss_dm = max(pow_gau-powinterp);
    max_power_loss_b = max_power_loss_dm; #WCE-DM and WCE-B have the same power property

    # --- Collect results
    if (verbose) {
      print(paste0("M = ", iM, " / ", m_set_length));
    }
    mat_size_distortion_dm[iM] = size_distortion_dm;
    mat_power_loss_dm[iM] = max_power_loss_dm;

    mat_size_distortion_b[iM] = size_distortion_b;
    mat_power_loss_b[iM] = max_power_loss_b;

  } #end of iM iteration

  df_hypoth_testing <- data.frame(
    v_M,
    v_hypothesis_test_b,
    v_test_statistic_b,
    v_hypothesis_test_dm,
    v_test_statistic_dm
  )

  df_size_power <- data.frame(
    M = m_set,
    b_size_distortion = mat_size_distortion_b[,1],
    b_power_loss = mat_power_loss_b[,1]
  )

  plotting_data <- merge(df_size_power, df_hypoth_testing, by.x = "M", by.y = "v_M")
  plotting_data["v_hypothesis_test_b"] <- ifelse(v_hypothesis_test_b == TRUE, "cross", "circle")


  plot <- ggplot(plotting_data, aes(x = .data$b_size_distortion, y = .data$b_power_loss), group=v_hypothesis_test_b) +
    geom_path(linewidth = 1, linetype = "dashed") +
    geom_point(aes(shape = v_hypothesis_test_b), size = 4.5, color = "red", stroke = 1.5) +
    labs(
      x = "Size Distortion",
      y = "Maximum Power Loss"
    ) +
    theme_minimal() +
    theme(legend.position = "bottom", legend.title = element_blank()) +
    scale_shape_manual(values=c(16, 4), labels = c("H0 not rejected", "H0 rejected"))

  # add m_set annotation
  if (no_m_label == FALSE){
    plot <- plot + geom_text(aes(label = M), nudge_y = .005)
  }

  return(list(plot, plotting_data))
}
